### PR TITLE
CHG: [droid] Adapt to ndk15+ with unified headers

### PR DESCRIPTION
**Pull request by koying**
As the name says, those are changes to the android build chain so as to adapt to the recent NDK move to an "unified headers" approach, ie. the headers are common to all toolchain/api, vs. headers specific per arch/api as it was before.

Big advantage: large offsets are now (more-or-less) properly handled, so no need to fiddle with off_t/off64_t anymore

### CODE REVIEW

File: tools/depends/configure.ac [hunk 1]
```diff
@@ -83,6 +83,11 @@ AC_ARG_WITH([sdk],
   [specify sdk platform version (optional for android). default is android-21])],
   [use_sdk=$withval])
 
+AC_ARG_WITH([ndk],
```
Comment (wsnipex) [inline tools/depends/configure.ac]: This changes behavior for an existing setting, should we go with api or ndk-version/level instead?

File: tools/depends/configure.ac [hunk 2]
```diff
@@ -83,6 +83,11 @@ AC_ARG_WITH([sdk],
...
+  [AS_HELP_STRING([--with-ndk],
+  [spcify ndk level (optional for android). default is 21])],
```
Comment (wsnipex) [inline tools/depends/configure.ac]: specify

File: tools/depends/configure.ac (hunk 1 above)
Comment (koying) [inline tools/depends/configure.ac]: OTOH it's now symetric with the sdk one.
If you want, we can have ndk-api and sdk-api, which makes sense.

Comment (wsnipex) [inline tools/depends/configure.ac]: yep, ndk-api and sdk-api sound sane

**Comment by peak3d** [comment]
Can we still build with older ndk (14e for example)?

**Comment by koying** [comment]
Yep, at least the last time I tried ;)

**Comment by peak3d** [comment]
Ok, then I'll give it a try here

**Comment by koying** [comment]
If by "here" you mean jenkins, that won't work as you cannot select the toolchain.

**Comment by peak3d** [comment]
here on my private system

File: tools/depends/configure.ac [hunk 3]
```diff
@@ -524,12 +532,12 @@ if test -z $use_tarballs; then
...
 if test "$platform_os" == "android"; then
-  if test -z $use_ndk; then
+  if test -z $use_ndk_path; then
     AC_MSG_ERROR("NDK path is required for android")
   fi
 
-  if ([! test -f "$use_ndk/source.properties"] && [! test -f "$use_ndk/RELEASE.TXT"]) ; then
-    AC_MSG_ERROR("$use_ndk is not an NDK directory")
+  if ([! test -f "$use_ndk_path/source.properties"] && [! test -f "$use_ndk_path/RELEASE.TXT"]) ; then
+    AC_MSG_ERROR("$use_ndki_path is not an NDK directory")
```
Comment (Memphiz) [inline tools/depends/configure.ac]: Typo ndki vs ndk

**Comment by Memphiz** [comment]
Not sure what to review exactly - those depends changes are android only or not?

File: tools/depends/configure.ac [hunk 4]
```diff
@@ -524,12 +532,12 @@ if test -z $use_tarballs; then
 fi
 
 if test "$platform_os" == "android"; then
-  if test -z $use_ndk; then
+  if test -z $use_ndk_path; then
     AC_MSG_ERROR("NDK path is required for android")
```
Comment (peak3d) [inline tools/depends/configure.ac]: would more like AC_MSG_ERROR("--with-ndk-path is required for android")

**Review by peak3d — CHANGES_REQUESTED** [review]

**Comment by koying** [comment]
@Memphiz Yes, they are *supposed* to be android-only, but as they touch depends, I prefer to ping you ;)

**Comment by Memphiz** [comment]
Any runtime checks to carry out with this then?

**Comment by koying** [comment]
@Memphiz Nothing specific. If it builds for ios/osx, it should be ok.
You can do a generic runtime test if you have time, but I guess a nightly post-merge would do.

**Review by Rechi — CHANGES_REQUESTED** [review]
rename `--with-ndk` to `--with-ndk-path` is missing in docs/README.android and tools/depends/README
samba fix should be moved ahead of default toolchain bump

**Comment by koying** [comment]
jenkins build this please

**Comment by koying** [comment]
https://github.com/android-ndk/ndk/issues/480

I think we can close this one. GCC is really getting end-of-life in NDK.
Next step should be to move to LLVM / libc++, really, if possible.

This patch was intended for easier migration to samba 4.5.1, but it seems there is no consensus on it, so bothering with this one becomes rather pointless

**Comment by MartijnKaijser** [comment]
you could still merge the ndk/sdk split

**Comment by koying** [comment]
Probably, yes. Will wait to see if it's relevant to clan, though.

The PIL one is definitely something we want, though.
Currently, we build with the native python options (via sysconfig), so it's a semi-miracle it works at all ;)
e.g., from random arm32 jenkins build:
```
/usr/bin/ccache /home/jenkins/android-dev//android-toolchain-arm-21-r14b/bin/arm-linux-androideabi-gcc -DANDROID -fexceptions -fPIC -DPIC -march=armv7-a -mtune=cortex-a9 -mfloat-abi=softfp -mfpu=neon -Og -g -D_DEBUG -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/android-21 -isystem /home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include -fPIC -fno-strict-aliasing -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include -DNDEBUG -g -O3 -Wall -Wstrict-prototypes -DHAVE_LIBJPEG -DHAVE_LIBZ -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/freetype2 -I/home/jenkins/workspace/Android-ARM/tools/depends/target/pythonmodule-pil/arm-linux-androideabi-android-21-debug/libImaging -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/arm-linux-androideabi-android-21-debug/include/python2.7 -I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include/python2.7 -c _imaging.c -o build/temp.linux-x86_64-2.7/_imaging.o
```

Some of the options (eg `-I/home/jenkins/workspace/Android-ARM/tools/depends/xbmc-depends/x86_64-linux-gnu-native/include`) come from `http://jenkins.kodi.tv/job/Android-ARM/ws/tools/depends/xbmc-depends/x86_64-linux-gnu-native/lib/python2.7/_sysconfigdata.py`

Note the **_x86_64-linux-gnu-native_** ...

**Comment by Rechi** [comment]
@koying this needs a rebase

**Comment by koying** [comment]
Abandonned

**Pull request closed without merging**